File: src/CsharpClient/QuixStreams.State/Storage/IStateStorage.cs
using System;
using System.Threading.Tasks;

namespace QuixStreams.State.Storage
{
    /// <summary>
    /// The minimum definition for a state storage
    /// </summary>
    public interface IStateStorage: IDisposable
    {
        /// <summary>
        /// Save raw data into the key
        /// </summary>
        /// <param name="key">Key of the element</param>
        /// <param name="data">Raw byte[] representation of data</param>
        /// <returns>Awaitable task</returns>
        public Task SaveRaw(string key, byte[] data);

        /// <summary>
        /// Load raw data from the key
        /// </summary>
        /// <param name="key">Key of the element</param>
        /// <returns>Awaitable result for raw byte[] representation of data</returns>
        public Task<byte[]> LoadRaw(string key);

        /// <summary>
        /// Remove key from the storage
        /// </summary>
        /// <param name="key">Key of the element</param>
        /// <returns>Awaitable task</returns>
        public Task RemoveAsync(string key);

        /// <summary>
        /// Check if storage contains key
        /// </summary>
        /// <param name="key">Key of the element</param>
        /// <returns>Awaitable result for boolean representing whether the storage contains key</returns>
        public Task<bool> ContainsKeyAsync(string key);

        /// <summary>
        /// Get list of all keys in the storage
        /// This function is written in the asynchronous manner
        /// </summary>
        /// <returns>Awaitable result for the keys as a hash set</returns>
        public Task<string[]> GetAllKeysAsync();

        /// <summary>
        /// Clear the storage / remove all keys from the storage
        /// This function is written in the asynchronous manner and returns Task
        /// </summary>
        /// <returns>Awaitable task</returns>
        public Task ClearAsync();

        /// <summary>
        /// Returns the number of keys in storage
        /// </summary>
        /// <returns></returns>
        public Task<int> Count();
        
        /// <summary>
        /// Returns whether the storage is case-sensitive
        /// </summary>
        public bool IsCaseSensitive { get; }
        
        /// <summary>
        /// Returns whether the transactions are supported
        /// </summary>
        public bool CanPerformTransactions { get; }
        
        /// <summary>
        /// Flush data to the storage
        /// </summary>
        /// <exception>Throws exception if the transaction fails</exception>
        public void Flush();
    }
    
    /// <summary>
    /// State storage types
    /// </summary>
    public enum StateStorageTypes
    {
        /// <summary>
        /// RocksDB storage
        /// </summary>
        RocksDb,
        
        /// <summary>
        /// In-memory storage
        /// </summary>
        InMemory
    }
}

File: src/PythonClient/src/quixstreams/statestorages/istatestorage.py
from abc import ABC, abstractmethod
from typing import List, Any, Optional

from .statevalue import StateValue
from .serializers.bytevalueserializer import ByteValueSerializer


class IStateStorage(ABC):
    """The minimum definition for a state storage"""

    @abstractmethod
    def save_raw(self, key: str, data: bytes) -> None:
        """Save raw data into the key"""
        ...

    @abstractmethod
    def load_raw(self, key: str) -> bytes:
        """Load raw data from the key"""
        ...

    @abstractmethod
    def remove(self, key: str) -> None:
        """Remove key from the storage"""
        ...

    @abstractmethod
    def contains_key(self, key: str) -> bool:
        """Check if storage contains key"""
        ...

    @abstractmethod
    def get_all_keys(self) -> List[str]:
        """Get list of all keys in the storage"""
        ...

    @abstractmethod
    def clear(self) -> None:
        """Clear the storage / remove all keys from the storage"""
        ...

    @abstractmethod
    def count(self) -> int:
        """Returns the number of keys in storage"""
        ...

    @property
    @abstractmethod
    def is_case_sensitive(self) -> bool:
        """Returns whether the storage is case-sensitive"""
        ...

    @abstractmethod
    def get_or_create_sub_storage(self, sub_storage_name: str, db_name: Optional[str] = None) -> 'IStateStorage':
        """Gets an existing sub-storage with the specified name or creates a new one if it does not already exist.

        Args:
            sub_storage_name: The name of the sub-storage to retrieve or create.
            db_name: The name of the database under which the storage will be created.
                     If this parameter is not specified, the storage will be created under the parent's database.

        Returns:
            IStateStorage: The state storage associated with the given sub-storage name.

        Raises:
            ValueError: Thrown when sub_storage_name is null or empty.
        """
        ...

    @abstractmethod
    def delete_sub_storage(self, sub_storage_name: str) -> bool:
        """Deletes a storage under this in hierarchy."""
        ...

    @abstractmethod
    def delete_sub_storages(self) -> int:
        """Deletes the storages under this in hierarchy."""
        ...

    @abstractmethod
    def get_sub_storages(self) -> List[str]:
        """Gets the storages under this in hierarchy."""
        ...

    @property
    @abstractmethod
    def can_perform_transactions(self) -> bool:
        """Return True if transactions are supported, False otherwise."""
        ...

    def start_transaction(self) -> None:
        """Starts a transaction."""
        ...

    def commit_transaction(self) -> None:
        """Commits a transaction.

        Raises:
            Exception: Thrown if the transaction fails.
        """
        ...

    def set(self, key: str, value: Any) -> None:
        """Save value into the key"""
        state_value = value if isinstance(value, StateValue) else StateValue(value)
        self.save_raw(key, ByteValueSerializer.serialize(state_value))

    def get(self, key: str) -> Any:
        """Load value from the key"""
        return ByteValueSerializer.deserialize(self.load_raw(key)).value
